[DROOLS-5517] Enhance DMNResult/DMNContext to be able to return stron…

tkobayas opened the pull request:
…gly typed OutputSet

WIP: This is a draft PR. Feel free to share your thoughts/ideas. @tarilabs @lucamolteni 

* I created DMNResultFPAImpl which has `getOutputSet()` method. But it means you need to cast the result in order to use the method. Should I move the method to Interface DMNResult?
* In this draft, DMNResultFPAImpl receives an OutputSet instance from DMNContextFPAImpl. It means that you need to instantiate InputSet/OutputSet instances before executing DMNRuntime. (as we do in BaseVariantTest.createTypeSafeInput/evaluateTypeSafe). Or do we want to encapsulate those logic inside DMNRuntime?

tarilabs commented on kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/DMNContextFPAImpl.java:
```diff
@@ -29,12 +29,20 @@
 
-    private final FEELPropertyAccessible fpa;
+    private final FEELPropertyAccessible inputSet;
+    private final FEELPropertyAccessible outputSet;
```
I am not sure we need for a context to hold reference for an input Vs output, given this would lead to a different design and semantic.

I believe it should be possible to keep `DMNContextFPAImpl` as is.

I also believe it should be possible to keep `DMNResultImpl` without any need for an additional specialization, but this less of a concern to me.

When the `DMNRuntimeImpl` creates the `DMNResultImpl`, it should be possible to supply a `DMNContext` which has for the _wrapped_ `FEELPropertyAccessible fpa` the full list of inputs,decisions,bkms, etc.

What do you think?

tkobayas commented:
Thank you for the review, @tarilabs 

> I believe it should be possible to keep DMNContextFPAImpl as is.
> 
> I also believe it should be possible to keep DMNResultImpl without any need for an additional specialization, but this less of a concern to me.
>
> When the DMNRuntimeImpl creates the DMNResultImpl, it should be possible to supply a DMNContext which has for the wrapped FEELPropertyAccessible fpa the full list of inputs,decisions,bkms, etc.

I updated the PR. This time, I keep `DMNContextFPAImpl` and `DMNResultImpl` unchanged. Instead, I introduced `DMNContextTypeSafeImpl` which works like a usual `DMNContextImpl` during runtime execution and also holds `fpa` (= `OutputSet`).

One main concern is how `DMNContextTypeSafeImpl` can know `OutputSet` class in order to convert the result map to the `fpa`. At the moment, `InputSet` and `OutputSet` are compiled/instantiated outside of `DMNRuntime` (e.g. `BaseVariantTest.evaluateTypeSafe()`). So in this PR, I supplied `OutputSet` class from the test case after `runtime.evaluateAll()` (I demonstrate it only in `BaseVariantTest.testOneOfEachType()` just for explanation). But probably it's not what you expect. Do you want to supply the `OutputSet` class when we call `evaluateAll()`? Then, how to supply it... (maybe using `DMNContextFPAImpl.metadata`?). I may still be misunderstanding your points. Feel free to share your thoughts. Thanks!

tkobayas commented on kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/stronglytyped/DMNRuntimeTypesTest.java:
```diff
@@ -98,7 +99,8 @@ public void testOneOfEachType() {
 
         if (isTypeSafe()) {
-            FEELPropertyAccessible outputSet = convertToOutputSet(dmnModel, dmnResult);
+            DMNContextTypeSafeImpl resultContext = (DMNContextTypeSafeImpl)dmnResult.getContext();
+            FEELPropertyAccessible outputSet = resultContext.getFpa(getOutputSetClass(dmnModel));
```
Here I supply `OutputSet` class.

tkobayas commented:
Jenkins trigger cdb

tkobayas commented:
Rebased because of conflict

tarilabs commented on kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/impl/DMNContextFPAImpl.java:
```diff
@@ -93,7 +149,15 @@ public DMNMetadata getMetadata() {
+                    throw new DMNTypeSafeException("Exception while instantiating " + fpa.getClass(), e);
+        }
+        newFpa.fromMap(fpa.allFEELProperties());
+        DMNContextFPAImpl newCtx = new DMNContextFPAImpl(newFpa, metadata.asMap(), fallbackMap);
```
```suggestion
        DMNContextFPAImpl newCtx = new DMNContextFPAImpl(newFpa, metadata.asMap(), fallbackMap.clone());
```

I assume?

tkobayas commented:
Rebased because of conflict. Also squashed commits because it would be easier to review.

tkobayas commented:
CDB error doesn't seem to be related:
```
17:36:00  [ERROR] Failed to execute goal on project google-calendar-workitem: Could not resolve dependencies for project org.jbpm.contrib:google-calendar-workitem:jar:7.43.0-SNAPSHOT: Could not transfer artifact com.google.oauth-client:google-oauth-client:jar:1.23.0 from/to mirror-central (https://bxms-qe.rhev-ci-vms.eng.rdu2.redhat.com:8443/nexus/content/groups/kie-all/): /home/jenkins/.m2/repository/com/google/oauth-client/google-oauth-client/1.23.0/google-oauth-client-1.23.0.jar.part (No such file or directory) -> [Help 1]
17:36:00  org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal on project google-calendar-workitem: Could not resolve dependencies for project org.jbpm.contrib:google-calendar-workitem:jar:7.43.0-SNAPSHOT: Could not transfer artifact com.google.oauth-client:google-oauth-client:jar:1.23.0 from/to mirror-central (https://bxms-qe.rhev-ci-vms.eng.rdu2.redhat.com:8443/nexus/content/groups/kie-all/): /home/jenkins/.m2/repository/com/google/oauth-client/google-oauth-client/1.23.0/google-oauth-client-1.23.0.jar.part (No such file or directory)
```

tarilabs reviewed (commented):
@tkobayas thank you!

Overall is fine, but I believe we should seize the chance to fully update as discussed in email thread, so that `DMNRuntimeImpl` is not responsible in having to deal with the classname for outputset.

I believe for now, we should let `DMNRuntimeImpl` just use internally `DMNContextImpl` for evaluation, and just return that out of `evaluate...` methods.

Then in the tests, since we are in full control of the name of the outputset class, we return that to the concrete test method. In other words the transformation from `DMNContext` based on map, to `DMNContext` based on the outputset FPA, is handled by the test infrastructure.

This way, we don't have to use metadata for the outputset classname, and `DMNRuntime` does not need any modification for that transformation.
I believe this will reflect more closely what we are meaning to do for REST enpoint on Kogito anyway.

What do you think?

tkobayas commented:
Jenkins retest this

tkobayas commented:
Thank you for reviewing, @tarilabs 

Yes, sure. I moved the conversion from DMNContextImpl to DMNContextFPAImpl to the test code. Now DMNRuntimeImpl is much simpler.

Is this good to merge? :)

tkobayas commented:
Jenkins run fdb

tarilabs commented on kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/typesafe/AbstractDMNSetType.java:
```diff
@@ -32,7 +32,8 @@
 
-    List<DMNDeclaredField> fields = new ArrayList<>();
+    List<DMNDeclaredField> dmnFields = new ArrayList<>();
+    List<FieldDefinition> extraFields = new ArrayList<>();
```
just to be sure I understood the latest modifications. This `extraFields` is used to store the `DefinedKeySetField` which is a `Set<String>` of the properties which have been set on the Input/Output Set pojo. However, this is not really "used" so far, since DMNRuntimeImpl uses the Map-backed DMNContextImpl. Correct?
It's fine I just wanted to check the intention :)

tkobayas replied:
@tarilabs 

`DefinedKeySetField` is actually used now. When we start strongly typed DMN, we instantiate `DMNContextFPAImpl` with an `OutputSet` instance. Then, `DMNContextFPAImpl` is converted to `DMNContextImpl` for evaluation. At this point, we call `fpa.allFEELProperties()` inside `DMNContextImpl.clone()` to convert the FPA fields to a Map. Here, we need to distinguish between "undefined" (= don't put in the Map) and "explicit null" (= put the key with null value). So `DefinedKeySetField` mechanism is required.

If we use `InputSet` for the first `DMNContextFPAImpl`, we don't need this mechanism because `InputSet` doesn't have Decision node fields so we can always assume "null" is "explicit null". So... if we decide back to the approach "Instantiate `DMNContextFPAImpl` with an `InputSet` instance first. Then convert the result to `OutputSet`", I can remove the `DefinedKeySetField`. WDYT?

tarilabs replied:
oh I see now, thanks @tkobayas for the explanation, makes sense.

I would say please leave this as-is.
This can be helpful for supporting the use case when for instance on Kogito, we want to supply pre-evaluated-decision via kogito scaffolding.

Given inputset and outputset are infrastructural classes and emerging by the design of the DMN model itself, I believe it is good we have this utility method to distinguish the cases, as you pointed out. I just wanted to be sure if it was actually used, and now I see how :) 

Thanks!

tarilabs reviewed (approved):
Looks good to me, thanks!
Just a question to check I understood correctly.

tarilabs commented:
![image](https://user-images.githubusercontent.com/1699252/92566596-c0a31080-f27c-11ea-87ba-c56c6646759b.png)

🤷‍♂️ 

Jenkins run cdb

lucamolteni commented:
> ![image](https://user-images.githubusercontent.com/1699252/92566596-c0a31080-f27c-11ea-87ba-c56c6646759b.png)
> 
> 🤷‍♂️
> 
> Jenkins run cdb

AFAIK 11 is borken

tarilabs commented:
@lucamolteni as you can see also in the screenshot, I re-triggered the C-D-B

tarilabs commented:
as most CDB are unrelated also on other PRs I would just leave as-is
```
11:14:14  [ERROR] Failed to execute goal on project google-drive-workitem: Could not resolve dependencies for project org.jbpm.contrib:google-drive-workitem:jar:7.44.0-SNAPSHOT: Could not transfer artifact com.google.oauth-client:google-oauth-client-java6:jar:1.23.0 from/to mirror-central (https://bxms-qe.rhev-ci-vms.eng.rdu2.redhat.com:8443/nexus/content/groups/kie-all/): /home/jenkins/.m2/repository/com/google/oauth-client/google-oauth-client-java6/1.23.0/google-oauth-client-java6-1.23.0.jar.part (No such file or directory) -> [Help 1]
11:14:14  org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal on project google-drive-workitem: Could not resolve dependencies for project org.jbpm.contrib:google-drive-workitem:jar:7.44.0-SNAPSHOT: Could not transfer artifact com.google.oauth-client:google-oauth-client-java6:jar:1.23.0 from/to mirror-central (https://bxms-qe.rhev-ci-vms.eng.rdu2.redhat.com:8443/nexus/content/groups/kie-all/): /home/jenkins/.m2/repository/com/google/oauth-client/google-oauth-client-java6/1.23.0/google-oauth-client-java6-1.23.0.jar.part (No such file or directory)
```
